lib/hostip.c

```diff
@@ -819,48 +819,95 @@ CURLcode Curl_loadhostpairs(struct Curl_easy *data)
     }
     else {
       struct Curl_dns_entry *dns;
-      Curl_addrinfo *addr;
+      Curl_addrinfo *head = 0, *tail = 0;
       char *entry_id;
       size_t entry_len;
-      char buffer[256];
-      char *address = &buffer[0];
-
-      if(3 != sscanf(hostp->data, "%255[^:]:%d:%255s", hostname, &port,
-                     address)) {
-        infof(data, "Couldn't parse CURLOPT_RESOLVE entry '%s'!\n",
-              hostp->data);
-        continue;
+      char address[64];
+      char *addrPointer;
+      char *last;
+      char *tmp;
+      unsigned long tmpPort;
+      int error = true;
+
+      tmp = strchr(hostp->data, ':');
+      if(!tmp || (tmp - hostp->data) >= (ptrdiff_t)sizeof(hostname)) {
+        goto err;
       }
+      memcpy(hostname, hostp->data, tmp - hostp->data);
+      hostname[tmp - hostp->data] = '\0';
 
-      /* allow IP(v6) address within [brackets] */
-      if(address[0] == '[') {
-        size_t alen = strlen(address);
-        if(address[alen-1] != ']')
-          /* it needs to also end with ] to be valid */
-          continue;
-        address[alen-1] = 0; /* zero terminate there */
-        address++; /* pass the open bracket */
+      tmpPort = strtoul(tmp + 1, &tmp, 10);
+      if(!tmpPort || tmpPort > USHRT_MAX || *tmp != ':' || !(tmp + 1)) {
+        goto err;
       }
+      port = (int)tmpPort;
+
+      while(true) {
+        size_t alen;
+
+        last = tmp + 1;
+        tmp = strchr(last, ',');
+        if(tmp)
+          alen = tmp - last;
+        else
+          alen = strlen(hostp->data) - (last - hostp->data);
+
+        if(alen >= sizeof(address))
+          goto err;
+
+        memcpy(address, last, alen);
+        address[alen] = '\0';
+
+        /* allow IP(v6) address within [brackets] */
+        if(address[0] == '[') {
+          if(address[alen-1] != ']')
+            /* it needs to also end with ] to be valid */
+            continue;
+          address[alen-1] = 0; /* zero terminate there */
+          addrPointer = address + 1; /* pass the open bracket */
+        }
+        else {
+          addrPointer = address;
+        }
+
+        if(tail) {
+          tail->ai_next = Curl_str2addr(addrPointer, port);
+          if(!tail->ai_next)
+            goto err;
+          tail = tail->ai_next;
+        }
+        else {
+          head = tail = Curl_str2addr(addrPointer, port);
+          if(!head)
+            goto err;
+        }
 
-      addr = Curl_str2addr(address, port);
-      if(!addr) {
-        infof(data, "Address in '%s' found illegal!\n", hostp->data);
+        if(tmp)
+          last = tmp + 1 + alen;
```

Comment:
this again I don't follow wouldn't it be out of bounds. if you checked for null it would be better like
if(*end)
begin = end;
else
break;

also it occurs to me i should have done this review in the aggregate. sorry for the slew of notifications!

Reply:
No worries :-)

Reply:
The + alen was incorrect. You're right. This would read out of bounds. I've fixed that now.